engines/plumbers/3do.cpp

```diff
@@ -0,0 +1,703 @@
+#include "graphics/surface.h"
+
+#include "image/cel_3do.h"
+
+#include "video/3do_decoder.h"
+
+namespace Plumbers {
+
+	// TODO(3do):
+// * effect when changing between scenes
+// * 3do boot logo
+// * return to previous scene
+// * hide cursor
+
+namespace {
+// TODO: discover correct offsets
+Common::Point getMikeStart(uint num, uint total) {
+	if (total == 2)
+		return Common::Point(140 * num + 10, 120 * num + 10);
+	return Common::Point(60 * num + 20, 70 * num + 20);
+}
+
+// TODO: discover correct offsets
+Common::Point getMikeSize(uint total) {
+	if (total == 2)
+		return Common::Point(80, 100);
+	return Common::Point(80, 60);
+}
+
+void makeMikeDecision(Scene &scene, uint num) {
+	scene._bitmapNum = 0;
+	scene._startBitmap = 0;
+	scene._decisionChoices = num;
+	scene._waveFilename = "";
+	scene._style = Scene::STYLE_DECISION_MIKE;
+
+	Common::Point sz = getMikeSize(num);
+
+	for (uint i = 0; i < num; i++) {
+		Common::Point ms = getMikeStart(i, num);
+		scene._choices[i]._region = Common::Rect(ms.x, ms.y, ms.x + 2 * sz.x, ms.y + sz.y);
+	}
+}
+
+void makeVideo(Scene &scene, const Common::String &videoName, const Common::String &nextScene) {
+	scene._bitmapNum = 0;
+	scene._startBitmap = 0;
+	scene._decisionChoices = 1;
+	scene._waveFilename = "";
+	scene._style = Scene::STYLE_VIDEO;
+	scene._sceneName = videoName;
+	scene._decisionBitmap = "";
+	scene._choices[0]._sceneName = nextScene;
+}
+
+const struct {
+	const char *from;
+	struct {
+		const char *scene;
+		int points;
+	} to[kMaxChoice];
+} tungraph[] = {
+	{
+		"dec13",
+		{
+			{"sc15", -10000},
+			{"sc17", -10000},
+			{"sc20", -50000}
+		}
+	},
+	{
+		"dec16",
+		{
+			{"dec13", -20000},
+			{"restart", 0}
+		}
+	},
+	{
+
+		"dec18",
+		{
+			{"dec13", -30000},
+			{"restart", 0}
+		}
+	},
+	{
+		"dec20",
+		{
+			{"sc21", -90000},
+			{"sc22", -90000}
+		}
+	},
+	{
+		"dec22",
+		{
+			{"dec20", -90000},
+			{"restart", 0}
+		}
+	},
+	{
+		"dec23",
+		{
+			{"sc24", 50000},
+			{"sc28", -50000}
+		}
+	},
+	{
+		"dec25",
+		{
+			{"sc26", -75000},
+			{"sc27", -90000}
+		}
+	},
+	{
+		"dec26",
+		{
+			{"dec25", -75000},
+			{"restart", 0}
+		}
+	},
+	{
+		"dec28",
+		{
+			{"dec23", -75000},
+			{"restart", 0}
+		}
+	},
+	{
+		"dec29",
+		{
+			{"sc30", -20000},
+			{"sc31", 90000}
+		}
+	},
+	{
+		"dec30",
+		{
+			{"sc32", 0},
+			{"restart", 0},
+			{"end", 0}
+		}
+	},
+	{
+		"dec31",
+		{
+			{"dec29", -20000},
+			{"end", 0}
+		}
+
+	}
+};
+
+static const Common::KeyCode cheatKbd[] = {
+	Common::KEYCODE_UP,
+	Common::KEYCODE_DOWN,
+	Common::KEYCODE_RIGHT,
+	Common::KEYCODE_LEFT,
+	Common::KEYCODE_DOWN,
+	Common::KEYCODE_RIGHT,
+	Common::KEYCODE_RETURN
+};
+
+static const Common::JoystickButton cheatJoy[] = {
+	Common::JOYSTICK_BUTTON_DPAD_UP,
+	Common::JOYSTICK_BUTTON_DPAD_DOWN,
+	Common::JOYSTICK_BUTTON_DPAD_RIGHT,
+	Common::JOYSTICK_BUTTON_DPAD_LEFT,
+	Common::JOYSTICK_BUTTON_DPAD_DOWN,
+	Common::JOYSTICK_BUTTON_DPAD_RIGHT,
+	Common::JOYSTICK_BUTTON_X
+};
+}
```

Comment:
I'd add here something like `// end of anonymous namespace`

Reply:
Done